pdns/dnsdistdist/dnsdist-opentelemetry.cc

```diff
@@ -0,0 +1,279 @@
+  return 0;
+#else
+  auto otTrace = pdns::trace::TracesData{
+    .resource_spans = {
+      pdns::trace::ResourceSpans{
+        .resource = {
+          .attributes = std::vector<pdns::trace::KeyValue>{
+            pdns::trace::KeyValue{
+              "service.name", pdns::trace::AnyValue{"dnsdist"}},
+          }},
+        .scope_spans = std::vector<pdns::trace::ScopeSpans>{{}}}}};
+
+  if (d_attributes != nullptr) {
+    otTrace.resource_spans.at(0).resource.attributes.insert(
+      otTrace.resource_spans.at(0).resource.attributes.end(),
+      d_attributes->begin(),
+      d_attributes->end());
+  }
+
+  if (d_preActivationSpans != nullptr) {
+    for (auto const& preActivationTrace : *d_preActivationSpans) {
+      otTrace.resource_spans.at(0).scope_spans.at(0).spans.push_back(
+        {
+          .trace_id = d_traceid,
+          .span_id = preActivationTrace.span_id,
+          .parent_span_id = preActivationTrace.parent_span_id,
+          .name = preActivationTrace.name,
+          .start_time_unix_nano = preActivationTrace.start_time_unix_nano,
+          .end_time_unix_nano = preActivationTrace.end_time_unix_nano,
+        });
+    }
+  }
+
+  if (d_postActivationSpans != nullptr) {
+    otTrace.resource_spans.at(0).scope_spans.at(0).spans.insert(
+      otTrace.resource_spans.at(0).scope_spans.at(0).spans.end(),
+      d_postActivationSpans->begin(),
+      d_postActivationSpans->end());
+  }
+  return otTrace;
+#endif
+}
+
+std::string Tracer::getOTProtobuf() const
+{
+#ifdef DISABLE_PROTOBUF
+  return 0;
+#else
+  // TODO: Should we close all spans?
+  return getTracesData().encode();
+#endif
+}
+
+SpanID Tracer::addSpan([[maybe_unused]] const std::string& name)
+{
+#ifdef DISABLE_PROTOBUF
+  return 0;
+#else
+  return addSpan(name, SpanID{});
+#endif
+}
+
+SpanID Tracer::addSpan([[maybe_unused]] const std::string& name, [[maybe_unused]] const SpanID& parentSpanID)
+{
+#ifdef DISABLE_PROTOBUF
+  return 0;
+#else
+  auto spanID = pdns::trace::randomSpanID();
+  if (d_activated) {
+    d_postActivationSpans->push_back({
+      .trace_id = d_traceid,
+      .span_id = spanID,
+      .parent_span_id = parentSpanID,
+      .name = name,
+      .start_time_unix_nano = pdns::trace::timestamp(),
+    });
+    return spanID;
+  }
+
+  // We're not activated, so we are in pre-activation.
+  if (d_preActivationSpans == nullptr) {
+    // This is the first time addEvent is called for this Tracer instance
+    d_preActivationSpans = std::make_unique<std::vector<preActivationSpanInfo>>();
+  }
+
+  d_preActivationSpans->push_back({
+    .name = name,
+    .span_id = spanID,
+    .parent_span_id = parentSpanID,
+    .start_time_unix_nano = pdns::trace::timestamp(),
+    .end_time_unix_nano = 0,
+  });
+
+  d_lastSpanID = spanID;
+
+  return spanID;
+#endif
+}
+
+// TODO: Figure out what to do with duplicate keys
+void Tracer::setTraceAttribute([[maybe_unused]] const std::string& key, [[maybe_unused]] const AnyValue& value)
```

Comment:
What's the impact of having duplicate keys? If we want to prevent them we can use a `unordered_map` instead of a `vector`, but I don't know if it is worth the trouble.